database/mysql/mysql.go

```diff
@@ -329,7 +324,7 @@ func (m *Mysql) SetVersion(version int, dirty bool) error {
 		return &database.Error{OrigErr: err, Err: "transaction start failed"}
 	}
 
-	query := "TRUNCATE `" + m.config.MigrationsTable + "`"
+	query := "DELETE FROM `" + m.config.MigrationsTable + "`"
```

Comment:
Did you also try a `DROP` + `CREATE`?

It seems like that was the [approach](https://github.com/golang-migrate/migrate/issues/584#issuecomment-867046189) MySQL took in v8. 

What are the differences in behavior and performance between `TRUNCATE` and `DELETE`?
Since this table should be small, I don't expect a significant performance difference.
I don't think `DELETE` will create a [table lock](https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html) where as `DROP` will. Also, it looks like neither `DROP` or `TRUNCATE` [can](https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html) be [rolledback](https://dev.mysql.com/doc/refman/8.0/en/cannot-roll-back.html).

Reply:
We should [centralize the discussion](https://github.com/golang-migrate/migrate/issues/584#issuecomment-867105762) :) . So yes, DELETE can be rolled back, and neither DROP nor TRUNCATE can as you mentioned. That's the main issue. The other issue is the performance of TRUNCATE in servers with a large buffer pool that affects 5.7 and below. 

Also, delete will create a lock over all the *rows* in the table, but not a *TABLE LOCK*, if that is what you need (lock the table), you could use [LOCK TABLES](https://dev.mysql.com/doc/refman/8.0/en/lock-tables.html) instead but I don't think that it is really needed because all other transactions will be waiting for the rowlock anyway.

Regards, 

Martin.

Reply:
Agreed, let's keep the discussion in the [issue](https://github.com/golang-migrate/migrate/issues/584).